### PR TITLE
Improve TorchScript module getattr() to be same as python class getattr() method

### PR DESCRIPTION
**Pull request by jerryztsai**
Issue: https://github.com/pytorch/pytorch/issues/56909

Note the emitted code for such a call will either be a) getattr() call with first two args if the
attribute name (which must be a string literal) is determined to be valid based on the hasAttr() result,
or b) just the AST node for the default value (the 3rd arg) alone with no getattr call at all.

Test code:

```
import torch
import numpy as np

class Shape:
    def __init__(self):
        self.center = 1.0

def f(x):
    s = Shape()
    return getattr(s, "missing", [])

y = torch.jit.script(f)
print(y.graph)
```
Output:
```
graph(%x : Tensor):
  %s.1 : __torch__.Shape = prim::CreateObject()
  %2 : NoneType = prim::CallMethod[name="__init__"](%s.1) # ts.py:10:8
  %4 : Tensor[] = prim::ListConstruct()
  return (%4)
```

Another example:
```
import torch

class Shape:
    def __init__(self):
        self.center = 1.0


def f(x):
    s = Shape()
    y = getattr(s, "center")
    w : list[float] = [1.0]
    z = getattr(s, "missing", w)
    z.append(y)
    return z


y = torch.jit.script(f)
print(y.graph)
--- output ---

graph(%x : Tensor):
  %5 : float = prim::Constant[value=1.]() # ts.py:12:23
  %s.1 : __torch__.Shape = prim::CreateObject()
  %2 : NoneType = prim::CallMethod[name="__init__"](%s.1) # ts.py:10:8
  %center : float = prim::GetAttr[name="center"](%s.1)
  %w.1 : float[] = prim::ListConstruct(%5)
  %11 : float[] = aten::append(%w.1, %center) # ts.py:14:4
  return (%w.1)
```
Fixes #{56969}

### CODE REVIEW

**Comment by jerryzhenleicai** [comment]
When I ran test, I got this error:

```

(torch) zcai@fractal:~/Dev/pytorch$ python test/test_jit.py 
Traceback (most recent call last):
  File "test/test_jit.py", line 7, in <module>
    from jit.test_tracer import TestTracer, TestMixTracingScripting  # noqa: F401
  File "/home/zcai/Dev/pytorch/test/jit/test_tracer.py", line 16, in <module>
    from torch.testing._internal.common_utils import suppress_warnings, \
  File "/home/zcai/Dev/pytorch/torch/testing/_internal/common_utils.py", line 46, in <module>
    import expecttest
ModuleNotFoundError: No module named 'expecttest'

```

File: requirements.txt [hunk 1]
```diff
@@ -9,3 +9,4 @@ setuptools
 six
 typing_extensions
 dataclasses; python_version<"3.7"
+expecttest
```
Comment (tugsbayasgalan) [inline requirements.txt:12]: Can you submit another PR for this? We normally try to have one PR per issue

File: test/jit/test_attr.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            s = Shape()
+            y = getattr(s, "center")
+            w : list[float] = [1.0]
+            z = getattr(s, "missing", w)
+            z.append(y)
+            return z
+
+
+        graph = torch.jit.script(func).graph
+        print(graph)
```
Comment (tugsbayasgalan) [inline test/jit/test_attr.py:37]: remove print statement

File: test/jit/test_attr.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+        # The "center" attribute is  valid on a shape object, so expect that
+        FileCheck().check("prim::GetAttr[name=\"center\"]").run(graph)
+        # The "missing" attribute does not exist, so there should be no corresponding GetAttr in AST
+        FileCheck().check_not("missing").run(graph)
+        # instead the getattr call will emit the default value, which is a list with one float element
+        FileCheck().check("float[] = prim::ListConstruct").run(graph)
```
Comment (tugsbayasgalan) [inline test/jit/test_attr.py:44]: is there a way to check how many elements are in the list? I think that would be better test

**Review by tugsbayasgalan — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 4]
```diff
@@ -2986,16 +3004,35 @@ struct to_ir {
         return std::make_shared<SimpleValue>(v);
       } break;
       case prim::GetAttr: {
-        checkApplyNumInputs(apply, 2);
+        checkApplyNumInputsRange(apply, 2, 3);
```
Comment (tugsbayasgalan) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3007]: nice

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 5]
```diff
@@ -2986,16 +3004,35 @@ struct to_ir {
...
         const std::string& name = StringLiteral(selector).text();
-        return obj->attr(apply.range(), method, name);
-      }
+        if (apply.inputs().size() == 2) {
+          return obj->attr(apply.range(), method, name);
+        } else {
+          // 3 inputs form of getattr, the third argument is the default value
+          // to return when attribute is not found
+          try {
+            return obj->attr(apply.range(), method, name);
```
Comment (tugsbayasgalan) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3021]: Can we just call hasattr to check if the attr is there? So you don't have to rely on the error type.

File: torch/csrc/jit/frontend/sugared_value.cpp [hunk 6]
```diff
@@ -209,6 +209,8 @@ std::shared_ptr<SugaredValue> SimpleValue::attr(
   ErrorReport report(loc);
   report << "'" << value_->type()->repr_str()
          << "' object has no attribute or method '" << field << "'.";
+  report.error_type = ErrorReport::ATTR_NOT_FOUND;
```
Comment (tugsbayasgalan) [inline torch/csrc/jit/frontend/sugared_value.cpp:212]: try to use hasattr so we don't need to change this logic

File: torch/csrc/jit/frontend/error_report.h [hunk 7]
```diff
@@ -12,6 +12,7 @@ struct Call {
 };
 
 struct TORCH_API ErrorReport : public std::exception {
+  enum { GENERIC, ATTR_NOT_FOUND } error_type = GENERIC;
```
Comment (tugsbayasgalan) [inline torch/csrc/jit/frontend/error_report.h:15]: unnecessary if hasattr works

File: test/jit/test_attr.py (hunk 3 above)
Comment (jerryzhenleicai) [reply test/jit/test_attr.py:34]: Now I verify that an append call is applied to the list in the IR (see line 36). Not clear how to check how many elements are exactly in the list.

File: requirements.txt (hunk 1 above)
Comment (jerryzhenleicai) [reply requirements.txt]: OK

File: test/jit/test_attr.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+class Shape:
+    def __init__(self):
+        self.center = 1.0
+
+
+class TestSetAttr(JitTestCase):
+    def test_setattr_with_default(self):
+        def func(x):
+            s = Shape()
```
[the remark this answers is not in the document]
Comment (jerryzhenleicai) [reply test/jit/test_attr.py]: now I use a module called A

File: test/jit/test_attr.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+                       "instead.")
+
+
+
+class Shape:
+    def __init__(self):
+        self.center = 1.0
+
+
+class TestSetAttr(JitTestCase):
```
[the remark this answers is not in the document]
Comment (jerryzhenleicai) [reply test/jit/test_attr.py]: ok

File: torch/csrc/jit/frontend/sugared_value.cpp (hunk 6 above)
Comment (jerryzhenleicai) [reply torch/csrc/jit/frontend/sugared_value.cpp]: done

File: test/jit/test_attr.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    def test_setattr_with_default(self):
+        def func(x):
+            s = Shape()
+            y = getattr(s, "center")
+            w : list[float] = [1.0]
+            z = getattr(s, "missing", w)
+            z.append(y)
+            return z
```
[the remark this answers is not in the document]
Comment (jerryzhenleicai) [reply test/jit/test_attr.py]: Added check for result having two elems. Thanks!

File: test/jit/test_attr.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        graph = torch.jit.script(A()).graph
+
+        # The "init_attr_val" attribute exists
+        FileCheck().check("prim::GetAttr[name=\"init_attr_val\"]").run(graph)
+        # The "missing" attribute does not exist, so there should be no corresponding GetAttr in AST
+        FileCheck().check_not("missing").run(graph)
+        # instead the getattr call will emit the default value, which is a list with one float element
+        FileCheck().check("float[] = prim::ListConstruct").run(graph)
+        # the IR for the .append() call before the return
+        FileCheck().check("float[] = aten::append(%w.1, %init_attr_val)").run(graph)
```
Comment (tugsbayasgalan) [inline test/jit/test_attr.py:38]: this feels little fragile because its too specific. (aka w.1) I wonder if this is even necessary?

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 12]
```diff
@@ -2901,6 +2901,24 @@ struct to_ir {
...
   }
 
+  void checkApplyNumInputsRange(Apply& apply, size_t min_expected_inputs,
+                                size_t max_expected_inputs) {
+    const SourceRange& loc = apply.range();
+    size_t position_arg_size = apply.inputs().size();
+    if (position_arg_size < min_expected_inputs ||
+        position_arg_size > max_expected_inputs) {
+      throw ErrorReport(loc)
+          << Var(apply.callee()).name().name() << " expected between "
```
Comment (tugsbayasgalan) [inline torch/csrc/jit/frontend/ir_emitter.cpp:2911]:
```suggestion
          << Var(apply.callee()).name().name() << " expected to have number of arguments between "
```

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 13]
```diff
@@ -2901,6 +2901,24 @@ struct to_ir {
...
+          << min_expected_inputs << " and " << max_expected_inputs
+          << " arguments but found "
```
Comment (tugsbayasgalan) [inline torch/csrc/jit/frontend/ir_emitter.cpp:2913]:
```suggestion
          << " but found "
```

**Review by tugsbayasgalan — APPROVED** [review]
